### ISSUE TITLE
MacroButton does not work with repeat parameters

### ISSUE THREAD

**Issue by stanislaw55** [sardana-org/sardana#1172]
Hi,
I've been trying to used sardana's `MacroButton` widget but with no success. 
This is the snippet which I used
```python
import sys

from taurus.qt.qtgui.application import TaurusApplication

app = TaurusApplication(sys.argv)
w = MacroButton()
w.setModel(sys.argv[1])
w.setMacroName("mv")
w.updateMacroArgument(0, "mot01")
w.updateMacroArgument(1, 100)
w.show()
sys.exit(app.exec_())
```

However, Door fails with error "Error: Moveable with name m does not exist"
Taurus version: 4.5.0
Sardana version: 2.7.1

It looks like this is error with Door itself. For some reason it splits string arguments and that causes error. However, I think this can come from wrong argument passing to `MacroButton`. Hence my question: how to properly use `MacroButton`

**Comment by reszelaz** [comment]
It may be a bug related to repeat parameters. `mv` macro uses repeat parameter.

Have you tried the main section of the macrobutton.py? There should be a widget with macro button and some text field for imputing parameters? Does it work with `mv`?

**Comment by stanislaw55** [comment]
Hi,
Another thing is that it works for macros without arguments, like `lsmac` of something similar.

> Have you tried the main section of the macrobutton.py?

Yes, I've tried that but there's only field for macro name and no place to provide macro arguments. See attached screenshot
![image](https://user-images.githubusercontent.com/32959446/61517933-b387ea00-aa08-11e9-880c-d27f134e4234.png)

When I tired to pass full macro with arguments (`mv mot01 100`) to macro name edit, it behaved just like I described earlier, resulting with error in Door.

So @reszelaz this may be the issue you pointed since running macro with limited number of arguments(`relmac ascan`) works fine. However, this seems like error in Door processing arguments, not MacroButton itself. Am I right?

**Comment by reszelaz** [comment]
`lsmac` has one repeat parameter called filter, and this is the text field that you see:
https://github.com/sardana-org/sardana/blob/9ea3b041491e776cd674e1b62700ede76037b536/src/sardana/macroserver/macros/lists.py#L49

I'm not sure but I think the limitation of the main section of the macrobutton.py module is that it only allows to pass one repetition of the repeat parameter. But this should not be the limitation of the `MacroButton` itself.

**Comment by stanislaw55** [comment]
So setting macro arguments by `updateMacroArgument` is correct approach?

**Comment by cpascual** [comment]
> So setting macro arguments by updateMacroArgument is correct approach?

Yes. Even if, admitedly, `setMacroArguments` would be a much more appropriate name for that method...

**Comment by stanislaw55** [comment]
@reszelaz , @cpascual  could you point me to place in codebase which may be responsible for such behaviour?

**Comment by stanislaw55** [comment]
Okay, it works when I change exact line to execute macro from
```python
self.door.runMacro(self.macro_name, self.macro_args)
```
to
```python
self.door.runMacro(self.macro_name, [[self.macro_args]])
```

**Comment by stanislaw55** [comment]
I've come up with this as a generic solution:
```python
    @ProtectTaurusMessageBox(msg='Error while executing the macro.')
    def runMacro(self):
        '''execute the macro with the current arguments'''
        if self.door is None:
            return
        param_defs = self.door.macro_server.getMacroInfoObj(self.macro_name).parameters
        parser = ParamParser(param_defs)
        parameters = parser.parse(" ".join(self.macro_args))
        try:
            self.door.runMacro(self.macro_name, parameters)
            sec_xml = self.door.getRunningXML()
            # get the id of the current running macro
            self.macro_id = sec_xml[0].get("id")
        except Exception, e:
            self.ui.button.setChecked(False)
            raise e
```

not the change in parsing parameters. Not using `shlex` anymore, just `sardana.util.parser.ParamParser`

**Comment by reszelaz** [comment]
Great! I confirm that there was a bug. With your solution this works correctly with `ascan`, `lsm` and `mv`.
Will you be so kind and prepare a PR? If this can wait, please prepare the PR after we merge #1173. I hope that next week we will be in Python 3 already.
Many thanks!

**Comment by stanislaw55** [comment]
You're welcome, @reszelaz . Actual it was @daneos who came up with idea of using `ParamParser`.
I'll create PR soon

**Comment by reszelaz** [comment]
Thanks to both! Again if this can wait, please propose a PR after migrating to Python 3. Cheers!

**Comment by stanislaw55** [comment]
I've found issue (maybe a bug?) with using MacroButton and Sequencer/MacroExecutor widget in one TaurusGui. For some reason, when I execute macro by MacroButton, it generates macro id with `uuid` module. MacorExecutor/Sequencer expects `int`.

I think this may be connected to passing macro name to MacroButton. When I use string with macro name, above error occures. May I should use different format?

@reszelaz could you confirm?

**Comment by reszelaz** [comment]
@stanislaw55 sorry for the delayed answer.

It would help a lot if you provide a simple example of such GUI (ideally with the sar_demo environment) so I could reproduce the problem. I will also ask my colleagues at ALBA if they use both widgets in the same GUI.

**Comment by stanislaw55** [comment]
I've been able to create simple GUI to recreate that issue. It turns out it is connected also to `sardana.taurus.qt.qtgui.macrolistener.MacroBroker` since its presence is required for error to occur.

The code snippet with example is presented below. It's almost 100 lines long but I've tried my best to reduce it as much as possible:
```python
from taurus.qt.qtgui.taurusgui.utils import PanelDescription

DOOR = "Door/local/1"
MS = "MacroServer/local/1"


macro_executor = PanelDescription(
    "Macro Executor",
    classname="TaurusMacroExecutorWidget",
    modulename="sardana.taurus.qt.qtgui.extra_macroexecutor.macroexecutor",
    model=MS,
    sharedDataRead={
        "macroserverName": "setModel",
        "doorName": "onDoorChanged",
        "macroStatus": "onMacroStatusUpdated",
    },
    sharedDataWrite={
        "macroName": "macroNameChanged",
        "executionStarted": "macroStarted",
        "shortMessage": "shortMessageEmitted",
    },
)

sequencer = PanelDescription(
    "Sequencer",
    classname="TaurusSequencerWidget",
    modulename="sardana.taurus.qt.qtgui.extra_macroexecutor.sequenceeditor.sequenceeditor",
    model=MS,
    sharedDataRead={
        "macroserverName": "setModel",
        "doorName": "onDoorChanged",
        "macroStatus": "onMacroStatusUpdated",
    },
    sharedDataWrite={
        "macroName": "macroNameChanged",
        "executionStarted": "macroStarted",
        "shortMessage": "shortMessageEmitted",
    },
)

from sardana.taurus.qt.qtgui.extra_macroexecutor.macrobutton import MacroButton
from taurus.qt.qtgui.dialog import ProtectTaurusMessageBox
from sardana.util.parser import ParamParser


class FixedMacroButton(MacroButton):

    @ProtectTaurusMessageBox(msg='Error while executing the macro.')
    def runMacro(self):
        """
        Overriden from MacroButton to parse macro parameters properly
        :return: None
        """
        if self.door is None:
            return

        param_defs = self.door.macro_server.getMacroInfoObj(self.macro_name).parameters
        parser = ParamParser(param_defs)
        parameters = parser.parse(" ".join(self.macro_args))

        try:
            self.door.runMacro(self.macro_name, parameters)
            sec_xml = self.door.getRunningXML()
            # get the id of the current running macro
            self.macro_id = sec_xml[0].get("id")
            print(self.macro_id)
        except Exception, e:
            self.ui.button.setChecked(False)
            raise e

if __name__ == '__main__':
    import sys

    from taurus.qt.qtgui.application import TaurusApplication
    from taurus.qt.qtgui.taurusgui import TaurusGui
    from taurus.external.qt import QtGui
    from taurus import Device

    from sardana.taurus.qt.qtgui.macrolistener import MacroBroker

    app = TaurusApplication(cmd_line_parser=None, app_name='MyGui')
    gui = TaurusGui(confname=__file__)
    gui.getPanel("Sequencer").widget().onDoorChanged(DOOR)
    gui.getPanel("Macro Executor").widget().onDoorChanged(DOOR)

    mb = FixedMacroButton()
    mb.setModel(DOOR)
    mb.setMacroName("wa")

    gui.createPanel(mb, "Macro Button")

    gui.__macroBroker = MacroBroker(gui)
    gui.__macroBroker.setModel(DOOR)

    gui.show()
    sys.exit(app.exec_())

```

**Comment by reszelaz** [comment]
Thanks @stanislaw55 for the example!
Unfortunatelly I did not have time to try it yet and I just leave for 3 weeks holidays.
I no one else take a look on it meanwhile I will after I come back:)

**Comment by reszelaz** [comment]
Hi Stanislaw,
I'm closing this issue since the original problem is solved. I would continue discussion about the macro id issue either on #1215 or directly on the PR that tries to solved it #1216.
